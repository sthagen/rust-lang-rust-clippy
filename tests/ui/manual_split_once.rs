// run-rustfix

#![feature(custom_inner_attributes)]
#![warn(clippy::manual_split_once)]
#![allow(clippy::iter_skip_next, clippy::iter_nth_zero)]

extern crate itertools;

#[allow(unused_imports)]
use itertools::Itertools;

fn main() {
    let _ = "key=value".splitn(2, '=').nth(2);
    let _ = "key=value".splitn(2, '=').nth(1).unwrap();
    let _ = "key=value".splitn(2, '=').skip(1).next().unwrap();
    let (_, _) = "key=value".splitn(2, '=').next_tuple().unwrap();

    let s = String::from("key=value");
    let _ = s.splitn(2, '=').nth(1).unwrap();

    let s = Box::<str>::from("key=value");
    let _ = s.splitn(2, '=').nth(1).unwrap();

    let s = &"key=value";
    let _ = s.splitn(2, '=').skip(1).next().unwrap();

    fn _f(s: &str) -> Option<&str> {
        let _ = s.splitn(2, '=').nth(1)?;
        let _ = s.splitn(2, '=').skip(1).next()?;
        let _ = s.rsplitn(2, '=').nth(1)?;
        let _ = s.rsplitn(2, '=').skip(1).next()?;
        None
    }

    // Don't lint, slices don't have `split_once`
    let _ = [0, 1, 2].splitn(2, |&x| x == 1).nth(1).unwrap();

    // `rsplitn` gives the results in the reverse order of `rsplit_once`
    let _ = "key=value".rsplitn(2, '=').nth(1).unwrap();
    let (_, _) = "key=value".rsplitn(2, '=').next_tuple().unwrap();
    let _ = s.rsplitn(2, '=').nth(1);
}

fn _msrv_1_51() {
    #![clippy::msrv = "1.51"]
    // `str::split_once` was stabilized in 1.52. Do not lint this
    let _ = "key=value".splitn(2, '=').nth(1).unwrap();
}

fn _msrv_1_52() {
    #![clippy::msrv = "1.52"]
    let _ = "key=value".splitn(2, '=').nth(1).unwrap();
}
