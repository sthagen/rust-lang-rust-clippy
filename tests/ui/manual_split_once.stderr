error: manual implementation of `split_once`
  --> $DIR/manual_split_once.rs:14:13
   |
LL |     let _ = "key=value".splitn(2, '=').nth(1).unwrap();
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `"key=value".split_once('=').unwrap().1`
   |
   = note: `-D clippy::manual-split-once` implied by `-D warnings`

error: manual implementation of `split_once`
  --> $DIR/manual_split_once.rs:15:13
   |
LL |     let _ = "key=value".splitn(2, '=').skip(1).next().unwrap();
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `"key=value".split_once('=').unwrap().1`

error: manual implementation of `split_once`
  --> $DIR/manual_split_once.rs:16:18
   |
LL |     let (_, _) = "key=value".splitn(2, '=').next_tuple().unwrap();
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `"key=value".split_once('=')`

error: manual implementation of `split_once`
  --> $DIR/manual_split_once.rs:19:13
   |
LL |     let _ = s.splitn(2, '=').nth(1).unwrap();
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `s.split_once('=').unwrap().1`

error: manual implementation of `split_once`
  --> $DIR/manual_split_once.rs:22:13
   |
LL |     let _ = s.splitn(2, '=').nth(1).unwrap();
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `s.split_once('=').unwrap().1`

error: manual implementation of `split_once`
  --> $DIR/manual_split_once.rs:25:13
   |
LL |     let _ = s.splitn(2, '=').skip(1).next().unwrap();
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `s.split_once('=').unwrap().1`

error: manual implementation of `split_once`
  --> $DIR/manual_split_once.rs:28:17
   |
LL |         let _ = s.splitn(2, '=').nth(1)?;
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `s.split_once('=')?.1`

error: manual implementation of `split_once`
  --> $DIR/manual_split_once.rs:29:17
   |
LL |         let _ = s.splitn(2, '=').skip(1).next()?;
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `s.split_once('=')?.1`

error: manual implementation of `rsplit_once`
  --> $DIR/manual_split_once.rs:30:17
   |
LL |         let _ = s.rsplitn(2, '=').nth(1)?;
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `s.rsplit_once('=')?.0`

error: manual implementation of `rsplit_once`
  --> $DIR/manual_split_once.rs:31:17
   |
LL |         let _ = s.rsplitn(2, '=').skip(1).next()?;
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `s.rsplit_once('=')?.0`

error: manual implementation of `rsplit_once`
  --> $DIR/manual_split_once.rs:39:13
   |
LL |     let _ = "key=value".rsplitn(2, '=').nth(1).unwrap();
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `"key=value".rsplit_once('=').unwrap().0`

error: manual implementation of `rsplit_once`
  --> $DIR/manual_split_once.rs:40:18
   |
LL |     let (_, _) = "key=value".rsplitn(2, '=').next_tuple().unwrap();
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `"key=value".rsplit_once('=').map(|(x, y)| (y, x))`

error: manual implementation of `rsplit_once`
  --> $DIR/manual_split_once.rs:41:13
   |
LL |     let _ = s.rsplitn(2, '=').nth(1);
   |             ^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `s.rsplit_once('=').map(|x| x.0)`

error: manual implementation of `split_once`
  --> $DIR/manual_split_once.rs:52:13
   |
LL |     let _ = "key=value".splitn(2, '=').nth(1).unwrap();
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `"key=value".split_once('=').unwrap().1`

error: aborting due to 14 previous errors

